File: problems/chapter06/miiitomi/003.cpp
// 第７回日本情報オリンピック 本選  C - ダーツ
// https://atcoder.jp/contests/joi2008ho/tasks/joi2008ho_c

#include <bits/stdc++.h>
using namespace std;

int main() {
    int N, M;
    cin >> N >> M;
    vector<int> P(N);
    for (int i = 0; i < N; i++) cin >> P[i];

    int K = ((N * (N + 1)) / 2) + N;
    vector<int> Q(K);
    int counter = 0;
    for (int i = 0; i < N; i++) {
        Q[counter] = P[i];
        counter++;
        for (int j = i; j < N; j++) {
            Q[counter] = P[i] + P[j];
            counter++;
        }
    }

    sort(Q.begin(), Q.end());

    int ans = 0;
    for (int i = 0; i < K; i++) {
        if (Q[i] > M) {
            break;
        }
        auto iter = lower_bound(Q.begin(), Q.end(), M - Q[i] + 1);
        if (iter == Q.begin()) {
            ans = max(ans, Q[i]);
        } else {
            iter--;
            ans = max(ans, Q[i] + *iter);
        }
    }

    cout << ans << endl;
}

File: problems/chapter06/mshibatatt/003.py
# https://atcoder.jp/contests/joi2008ho/tasks/joi2008ho_c
from bisect import bisect_right

def main():
    N, M = map(int, input().split())
    p = [int(input()) for _ in range(N)]
    
    # append 0 to consider skipping
    p_2 = [i+j for i in p for j in p+[0]] + [0]
    N_2 = N*(N+1)

    p_2.sort()
    output = -float('Inf')
    for i in p_2:
        if i > M:
            continue
        if i == M:
            output = M
            break
        target_score = M - i
        if target_score >= p_2[N_2]:
            index = N_2
        else:
            # binary search to find max key that p[key] =< target_score
            index = bisect_right(p_2, target_score) - 1
        output = max(output, i + p_2[index])

    print(output)

if __name__=='__main__':
    main()